### Luc√≠a de Lamadrid - Biotechnology (UPM)

This code was used to achieve the objectives of the Biotechnology Bachelor's thesis "TITLE." The methods were divided into six main stages: 
(1) Data acquisition, including sample preparation and sequencing.
(2) Alignment and quantification of reads with a reference genome.
(3) Differential expression analysis using DESeq2.
(4) Gene annotation through genomic databases.
(5) Gene Set Enrichment Analysis (GSEA).
(6) Application of machine learning methods for the selection and analysis of relevant variables.

![image](https://github.com/luciadlm/RNAseq/assets/172217433/769218b5-67cc-4f48-8ab2-272cef165e48)

Among the repository files, you will find alignment, feature counting, differential gene expression analysis (DGEanalysis.R), 
subsequent gene annotation (annotation.R), and GSEA (gsea.R). The files efa.R, ca.R, and featureselection.R correspond to 
learning methods that can be useful for interpreting differential expression data in a simplified manner, 
removing noise, and highlighting the most relevant genes.

